mantidproject/mantid
# Refactor ExtractSpectra to add MPI support

The refactor includes a series of additions and lower-level changes:

1. `ExtractSpectra` is doing two things, extracting spectra and cropping spectra. The "extract" part has been extracted into `ExtractSpectra2`.
2. `ExtractSpectra2` is using the "workspace property with index` mechanism added recently. It thus has a consistent interface for defining indices and also gives MPI support basically for free. `ExtractSpectra2` thus has different properties for defining indices and is also missing the X range properties, so its interface is not compatible with `ExtractSpectra`. Thus, we are currently not registering it as an algorithm since that would break scripts, but simply use it internally.
3. `ISpectrum::copyDataFrom` adds a mechanism for copying data from one `ISpectrum` to another with double dynamic dispatch. That is we can use it to copy data from one `MatrixWorkspace` to another without casting to `EventWorkspace` and implementing two different `exec()` versions for events and histograms.
4. Mechanism for extracting indices/grouping while preserving assignment of spectra to MPI ranks added to `SpectrumNumberTranslator` and `IndexInfo`.
5. Added Histogram slicing is currently visible in the diff but has been extracted into a separate PR so it can be ignored in this review: https://github.com/mantidproject/mantid/pull/21153.


**To test:**

Code review should be really thorough since `ExtractSpectra` was quite hard to see through and the refactor is not so trivial.

Fixes #21128.
No release notes, internal change.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
